pkg/sidecarexec/sandbox_cmd_runner_linux.go

```diff
@@ -0,0 +1,282 @@
+	}
+
+	wrapArgs := sandboxWrapArgs{
+		NewRootDir: newRootDir,
+		Posix:      r.opts.RequiresPosix != nil && r.opts.RequiresPosix[filepath.Base(cmd.Path)],
+		Network:    r.opts.RequiresNetwork != nil && r.opts.RequiresNetwork[filepath.Base(cmd.Path)],
+
+		CmdDir:  cmd.Dir,
+		CmdPath: cmd.Path,
+		CmdArgs: cmd.Args[1:], // drop first arg (binary name)
+
+		VisiblePaths: opts.VisiblePaths,
+	}
+
+	// Roughly equivalent to: unshare -Urm -R / ...
+	cmd.Path = "/usr/bin/kapp-controller"
+	cmd.Args = []string{"kapp-controller", "--sidecarexecwrap", wrapArgs.AsString()}
+	cmd.Dir = "/"
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		// TODO syscall.CLONE_NEWUTS | syscall.CLONE_NEWIPC | syscall.CLONE_NEWPID?
+		Cloneflags:  syscall.CLONE_NEWUSER | syscall.CLONE_NEWNS,
+		UidMappings: []syscall.SysProcIDMap{{ContainerID: 0, HostID: os.Geteuid(), Size: 1}},
+		GidMappings: []syscall.SysProcIDMap{{ContainerID: 0, HostID: os.Getegid(), Size: 1}},
+	}
+
+	// Create new non-configured network namespace if networking is not required
+	if !wrapArgs.Network {
+		cmd.SysProcAttr.Cloneflags = cmd.SysProcAttr.Cloneflags | syscall.CLONE_NEWNET
+	}
+
+	err = r.local.Run(cmd, opts)
+	if err != nil {
+		_ = os.RemoveAll(newRootDir) // remove error is lower priority
+		return err
+	}
+
+	// Propagate up any errors deleting root directory
+	// (want to notice any deletion problems asap)
+	err = os.RemoveAll(newRootDir)
+	if err != nil {
+		return fmt.Errorf("Deleting new root dir: %s", err)
+	}
+
+	return nil
+}
+
+// RunWithCancel executes exec.Cmd.
+// Kills execution immediately if value is read from cancelCh.
+func (SandboxCmdRunner) RunWithCancel(cmd *goexec.Cmd, cancelCh chan struct{}, opts exec.RunOpts) error {
+	return fmt.Errorf("RunWithCancel not implemented")
+}
+
+type sandboxWrapArgs struct {
+	NewRootDir string
+	Posix      bool
+	Network    bool
+
+	CmdDir  string
+	CmdPath string
+	CmdArgs []string
+
+	VisiblePaths []string
+}
+
+func newSandboxWrapArgs(str string) sandboxWrapArgs {
+	var args sandboxWrapArgs
+	err := json.Unmarshal([]byte(str), &args)
+	if err != nil {
+		panic(fmt.Sprintf("Internal inconsistency: invalid sandboxWrapArgs: %s", err))
+	}
+	return args
+}
+
+func (a sandboxWrapArgs) AsString() string {
+	bs, err := json.Marshal(a)
+	if err != nil {
+		panic(fmt.Sprintf("Internal inconsistency: invalid sandboxWrapArgs: %s", err))
+	}
+	return string(bs)
+}
+
+// SandboxWrap represents continuation of sandbox setup (mounting, ...)
+// which sandbox has initiated within a new user/mount namespace.
+type SandboxWrap struct{}
+
+type mountPoint struct {
+	Path      string
+	File      bool
+	Writeable bool
+	NonMount  bool
+	Content   string
+}
+
+// ExecuteCmd mounts various directories needed to execute sandboxed command.
+// As a last step it exec-s into given command, replacing the current process.
+func (w SandboxWrap) ExecuteCmd(rawArgs []string) error {
+	runtime.LockOSThread() // as a safety precaution for syscalls below
```

Comment:
I'm still groking the architecture here so i'm making a few assumptions...

There are some syscalls made below. mounting for e.g. 

If the syscalls below block, we may reach the max os threads causing our sandbox cmd runner to essentially freeze up / not process commands.

iirc GOMAXPROCS defaults to the number of CPUs, so there may not be many OS threads by default too.

I wonder if perhaps `runtime.LockOSThread() / runtime.UnLockOSThread()` should be done only during a syscall to minimize the time spent locked...

or

whether some timeouts should be introduced to the syscalls made.

Reply:
SandboxWrap class gets executed in a standalone process (called via kapp-controller --sidecarexecwrap flag) and the end of the function does an exec so nothing else should be executing within this process.

Reply:
Ah ok I see! Yeah I thought I was in the long running process. Getting a better handle on the arch now thx...